packages/plugin-ext/src/plugin/uri-ext.ts

```
@@ -0,0 +1,72 @@
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import * as theia from '@theia/plugin';
+import {
+    UriExt,
+    PLUGIN_RPC_CONTEXT, PluginInfo, UriMain
+} from '../common/plugin-api-rpc';
+import { RPCProtocol } from '../common/rpc-protocol';
+import { Disposable, URI } from './types-impl';
+import { UriComponents } from '../common/uri-components';
+
+export class UriExtImpl implements UriExt {
+
+    private static handle = 0;
+    private handles = new Set<string>();
+    private handlers = new Map<number, theia.UriHandler>();
+
+    private readonly proxy: UriMain;
+
+    constructor(readonly rpc: RPCProtocol) {
+        this.proxy = rpc.getProxy(PLUGIN_RPC_CONTEXT.URI_MAIN);
+        console.log(this.proxy);
+    }
+
+    registerUriHandler(handler: theia.UriHandler, plugin: PluginInfo): theia.Disposable {
+        const extensionId = plugin.id;
+        if (this.handles.has(extensionId)) {
+            throw new Error(`URI handler already registered for extension ${extensionId}`);
+        }
+
+        const handle = UriExtImpl.handle++;
+        this.handles.add(extensionId);
+        this.handlers.set(handle, handler);
+        this.proxy.$registerUriHandler(handle, extensionId, plugin.displayName || plugin.name);
+
+        return new Disposable(() => {
+            this.proxy.$unregisterUriHandler(handle);
+            this.handles.delete(extensionId);
+            this.handlers.delete(handle);
+        });
+    }
+
+    $handleExternalUri(handle: number, uri: UriComponents): Promise<void> {
+        const handler = this.handlers.get(handle);
+
+        if (!handler) {
+            return Promise.resolve(undefined);
+        }
+        try {
+            handler.handleUri(URI.revive(uri));
+        } catch (err) {
+            console.log(`error while handling external uri: ${uri}`);
```

Why not let the error propagate?